crates/acvm/src/pwg/binary.rs

```diff
@@ -0,0 +1,234 @@
+use std::collections::{BTreeMap, HashMap, HashSet};
+
+use acir::{
+    circuit::{gate::Directive, Gate},
+    native_types::{Expression, Witness},
+    FieldElement,
+};
+use num_bigint::BigUint;
+use num_traits::{One, Zero};
+
+use crate::GateResolution;
+
+pub struct BinarySolver {
+    binary_witness: HashSet<Witness>,
+    invert_witness: HashMap<Witness, Witness>,
+    positive_witness: HashMap<Witness, BigUint>,
+}
+
+impl Default for BinarySolver {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
```

Comment:
We should add documentation here explaining what it does and why it is needed

Reply:
I suppose you meant the BinarySolver, not the new() method. I added some comments.